util/buffer_received_can_message.h

```
@@ -0,0 +1,71 @@
+#ifndef BUFFER_RECEIVED_CAN_MESSAGE_H_
+#define BUFFER_RECEIVED_CAN_MESSAGE_H_
+
+/*
+ * A module for buffering CAN messages that are received from the CAN module.
+ * The module operates as a ring buffer, with the memory provided by the caller
+ * of the module.
+ *
+ * The whole goal of this module is allowing you to write application code that
+ * doesn't have to concern itself with ISR vs main thread code. If you set
+ * buffer_received_can_message as the can callback, then you can dequeue the
+ * buffered messages from the main thread without worrying about missing any
+ * messages from the bus.
+ *
+ * The ring buffer is designed to be pseudo-thread safe. None of the functions
+ * are rentrant, and there are no locking or atomic mechanisms, but flags are
+ * used to show whether a particular memory element contains valid data. This
+ * should allow the writer function (buffer_received_can_message) and the reader
+ * function (get_buffered_can_message) to operate from separate contexts (the
+ * former can be run in the ISR and the latter in the main thread) without any
+ * problems.
+ */
+
+#include "can.h"
+#include <stdbool.h>
+#include <stddef.h>
+
+/*
+ * Initializes the module. Pool is a memory buffer, which must be provided by
+ * the caller. pool_size is the size of pool, in bytes (not in number of
+ * can_t's)
+ */
+void receive_buffer_init(void *pool, size_t pool_size);
+
+/*
+ * Copies msg into our internal buffering system.
+ *
+ * This function fails silently if we're out of memory/space to hold the CAN
+ * message. This is so that you can use this buffering system as the CAN
+ * callback, whose signature is void
```

Does it at least set a flag? I know there's a "check if the buffer is full" function but that doesn't help if you're using this as the CAN callback as you're suggesting. Let's avoid silent failures.

---

I'll make it set a flag